top-level: simplify stdenv calling of top-level: fewer inferred and re-inferring arguments

Ericson2314 opened the pull request:
### New summery

Ok, this is last general purpose cleanup from me before going into the nitty-gritty details of cross compiling. The basic issue is despite my changes so far *almost* but not quite break the top-level stdenv cycle. This breaks the cycle, *and* cuts down on the partial application magic that can make it hard to figure out where an argument comes from.

This PR has been tested with my usual gist
```sh
nix-instantiate --eval -E 'import (builtins.fetchTarball https://gist.github.com/Ericson2314/fc745290db68b20235bdc457961b3658/archive/master.tar.gz) {}' -A testsPass
```
now extended to also check ghc's linux and darwin hashes in addition to cross + custom stdenv stuff. It should also pass travis--I ran the nox script locally.

In the process of cleaning up the default args used for testing, I added a builtins trace as they should never be triggered in production.

### Original Post

This grew out of the, ahem, tamer of my ideas in #10874. The basic idea is that `top-level/default.nix` is now only responsible for building a single bootstrapping phase, and since stdenvs control the bootstrapping, they are solely responsible for calling `top-level/default.nix` (well, with the exception of the final phase, since that isn't composed in a stdenv-specific manner). Additionally config's `replaceStdenv` and the cross stdenv are now composed more like other stdenvs than the ad-hoc way they were before.

While this doesn't really solve any problem on its own, it should lay a better foundation for feature work. See the comments in #14965 for how the phasing for the cross stdenv helps for example. Also, some of the more ambitious proposals of #10874 should similarly be assisted.

I tested (a few arbitrary) package paths stayed the same on Nixos. I verified the same for cross-compiling before I rebased onto the newly-bumped nixos-unstable, but post-rebase, cross-compiling seems to be broken for reasons unrelated to this.

The biggest problem with this is now the pkg hierarchy makes less sense: `top-level` isn't really the top-level and theirs a few more wrapper files. I'd gladly move things around (`phase-top-level`?) but not sure what the best way forward is.

I view this the spiritual successor of #14000, and as such, I tried to mimic @nbp's method of breaking this up into little commits to make it easier to review.

CC @vcunat @Mathnerd314 @nbp 

[PS: @nbp while you weren't involved in the #10874 discussion, this (hopefully just syntactically) conflicts a fair amount with @10874 hence the CC.]

domenkozar commented:
If we keep improving top-level functions (which I fully support), we should also document the flow so that we can reason about the improvements.

It's really hard to evaluate and follow the logic without spending considerable amount of time on it.

Ericson2314 commented:
Yeah I fully agree. I tried to add some documentation as I went along, but it would be good to move things into saner locations and add more.

Mathnerd314 commented on maintainers/scripts/dep-licenses.sh:
```diff
@@ -38,10 +38,10 @@ if test $(wc -l "$tmp/filter.sed" | sed 's/ .*//') == 0; then
-+  defaultStdenv = traceDrvLicenses allStdenvs.stdenv;
+-  defaultStdenv = vanillaStdenv;
++  defaultStdenv = traceDrvLicenses avanillaStdenv;
 "
```
Could this script be changed to use ``nix-instantiate -E``? Then this message would not be necessary.
Also, @nbp, is this script still useful?

Ericson2314 replied:
Technically, I didn't even fix the error message to be a real diff. I was mainly interested in making sure my `allStdenvs` grep came clean before I removed it :).

Mathnerd314 commented:
As far as the locations, I think ``pkgs/default.nix`` should be merged back into ``pkgs/top-level/default.nix`` (pure default arguments are usually helpful), and then ``pkgs/impure.nix`` can be moved to ``top-level`` as well, or else merged into the ``default.nix`` in the root directory (I can't say top-level, because that would be confusing).

Mathnerd314 commented on pkgs/top-level/default.nix:
```diff
@@ -19,83 +25,25 @@
                && system != "x86_64-solaris"
                && system != "x86_64-kfreebsd-gnu")
 
```
``noSysDirs`` should be in stdenv (true in stdenvNative and false elsewhere), rather than being a top-level argument.

Ericson2314 replied:
I like that idea, but I see it less as something that should have been done already than something that's good to do in addition---I haven't messed with any parameters that actual packages receive yet.

That said, I'm fine just making it an additional commit that's part of this rather than a separate PR.

Ericson2314 replied:
I have a similar idea for `crossSystem`. Now that `crossStdenv` is properly staged, we can give each stdenv a `host` and `target` system+platform, and set them accordingly in the cross stages.

`crossStdenv` would remain as a entry-point argument, but `all-packages.nix` would not have that argument as just look at `stdenv.host` and `stdenv.target`.

Ericson2314 replied:
But that, if not the `noSysDirs` change, will probably be more contentious and deserving of its own PR.

Mathnerd314 commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -4,10 +4,12 @@
-, allPackages ? import ../../..
-, platform ? null, config ? {}, lib ? (import ../../../lib)
+# Use stdenv parameter of `pkgs/top-level.nix` for easy testing.
+{ system
```
``pkgs/top-level.nix`` does not exist.

Ericson2314 replied:
Haha, fixed

Ericson2314 commented:
@Mathnerd314 `pkgs/default.nix` composes the whole thing, while `pkgs/top-level/default.nix` composes just one stage --- its not just a matter of default arguments.

How about:
 - `pkgs/top-level/default.nix`  -> `pkgs/top-level/stage.nix`
 - `pkgs/default.nix`  -> `pkgs/stage.nix`

I do like the idea of just sticking the impure stuff in `/defaut.nix` :)

Mathnerd314 commented:
> pkgs/default.nix composes the whole thing, while pkgs/top-level/default.nix composes just one stage --- it's not just a matter of default arguments.

I see, but that's still not a barrier to merging them. I did so [here](https://github.com/Ericson2314/nixpkgs/compare/cross...Mathnerd314:cross) ([full diff](https://github.com/NixOS/nixpkgs/compare/master...Mathnerd314:cross)), along with my other suggested changes.

Ericson2314 commented:
I like it! Part of me is still partial to a `default.nix` `phase.nix` separation, but it hardly matters. Note that @nbp purposely kept the throw-away-everything semantics for `pkgsWithOverrides` rather than using `__unfix__` for backwards compat, but I'm not at all sure that's preserving a feature rather than a bug. Thanks for taking a good look at this!

Ericson2314 commented:
Anybody else else want to review this (and @Mathnerd314's additions) soon? In some cases I imagine it might make sense to interleave/squash our commits---let me know where and I'll be happy to do it.

zimbatm commented on default.nix:
```diff
@@ -6,4 +6,4 @@ if ! builtins ? nixVersion || builtins.compareVersions requiredVersion builtins.
 else
 
-  import ./pkgs/top-level
+  import ./pkgs/impure.nix
```
I'm not sure about the location and name about this-one but don't have a better recommendation at the moment.

zimbatm commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -15,15 +16,14 @@
     cpio    = fetch { file = "cpio";  sha256 = "0ms5i9m1vdksj575sf1djwgm7zhnvfrrb44dxnfh9avr793rc2w4"; };
     tarball = fetch { file = "bootstrap-tools.cpio.bz2"; sha256 = "1lz1b0grl4642h6n635xvi6imf0yyy1zyzdr9ing5aphzz0z5iic"; executable = false; };
   }
+, ...
```
typo ?

Ericson2314 replied:
I don't think so. the `...`s allows us to give the all stdenvs the same arguments not worrying about which subset they actually use. See the simplifications to `stdenv/default.nix` in the same commit that introduced this.

zimbatm replied:
I see, makes sense

zimbatm commented:
Looking like a good cleanup

Ericson2314 commented:
Hello reviewers! :)

>  - `pkgs/top-level/default.nix`  -> `pkgs/top-level/stage.nix`
 - `pkgs/default.nix`  -> `pkgs/top-level/default.nix`

I rebased on top of latest channels/nixpkgs-unstable, did that rename I proposed, and added @Mathnerd314's commits to make it easier to review everything.

zimbatm commented:
I always wondered, why is that folder called top-level when it's not the top level? Having all the packages listen in `pkgs/default.nix` would mean less typing for me. And all the `<lang>-packages.nix` could go in `pkgs/development/<lang>-modules/default.nix`. Especially the python one who just references tons of patches from that folder anyways.

Ericson2314 commented:
@zimbatm I agree. I tried to conservatively keep things in the same place for now. But nixpkgs is overdue for a big sorting.

bjornfor commented on pkgs/top-level/default.nix:
```diff
@@ -50,9 +48,16 @@ let
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
-  };
+  # A few packages make make a new package set to draw their dependencies from.
```
"make make" => "make"

Ericson2314 replied:
Fixed, thanks!

zimbatm commented:
@Ericson2314 agreed not in this PR. Speaking of which, is it ready to be merged?

Ericson2314 commented:
@zimbatm Well, the first *n* commits are ready, the question is what *n* is :). I am quite confident that everything through my penultimate commit is a good idea. But I'm a bit unsure whether the last commit makes things more or less convoluted. And I am still triaging @Mathnerd314's commits: the guaranteed good ones I should probably rebase on top of my guaranteed good ones.

Ericson2314 commented:
Some decisions like "should `impure.nix` be part of `./default.nix`" don't affect complexity one way or the other and our purely stylistic. Whoever has authority to merge this should comment on those (I am guessing @edolstra, as this lies at the heart of nixpkgs).

zimbatm commented on default.nix:
```diff
@@ -7,33 +7,31 @@ if ! builtins ? nixVersion || builtins.compareVersions requiredVersion builtins.
+            let
+              toPath = builtins.toPath;
+              getEnv = x: if builtins ? getEnv then builtins.getEnv x else "";
+              pathExists = name:
```
toPath and pathExists are available since nix 0.10

Ericson2314 replied:
Fixed, thanks!

zimbatm replied:
Also see bec28d748c854abda373ca38831f4e77bc276fc1 in master

Ericson2314 commented:
Breaking this up to get things going with https://github.com/NixOS/nixpkgs/pull/16384 and https://github.com/NixOS/nixpkgs/pull/16382 being the first batch of trivial PRs.

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/16416 and https://github.com/NixOS/nixpkgs/pull/16415

Ericson2314 commented:
Ok, I finished triaging @mathnerd's commits. Everything is either already merged, squashed into my commits, or left to be dealt with after (see my `cross-rest` branch). I'm now trying to get travis to pass this---I fixed some legitimate bugs, and the current the current failure is due to somebody else's recent commit.

In light of all this, I am stripping the `semi-wip` prefix.

Ericson2314 commented:
I don't know how many people we need to review a change like this, but @vcunat, you mentioned you were interested at one point. Now would be a good time to take a look.

Ericson2314 commented:
Passes!

joachifm commented on pkgs/top-level/default.nix:
```diff
@@ -1,36 +1,43 @@
-   them with appropriate arguments.  The result is a set of all the
-   packages in the Nix Packages collection for some particular
-   platform. */
+/* This file composes a the Nix Packages collection. It:
```
"a the" looks like a typo

Ericson2314 replied:
Indeed it is. Thanks!

vcunat commented:
I've been severely out of time lately, and it's likely to last for weeks still. I'm sorry.

Ericson2314 commented:
That's ok. Thanks for saying so.

Ericson2314 commented:
@viric, @edolstra thought you might be a good person to review this?

[The merge conflict is trivial, but I am waiting for channels before rebasing so I can get a meaningful travis result.]

copumpkin commented:
I like where this is going 👍

You might also be interested in the related (though possibly more ambitious) https://github.com/NixOS/nixpkgs/issues/11486

Ericson2314 commented:
Normal packages do have their hashes preserved, but I'm getting missing `isFreeBSD` flags for some reason on `xburstutils`, one of the few (the only?) users of cross toolchains in normal uncross nixpkgs. 

@copumpkin the last commit is my attempt to make debugging stdenvs with this new system easier, but I'm worried it just over-complicates. As someone who's definitely put blood sweat and tears into debugging that :), what do you think?

Ericson2314 commented:
Ok fixed (i.e. preserved the weirdness of) cross compiling--this PR should not change the semantics of that.

Ericson2314 committed:
top-level: Stop exposing all stdenvs

Ericson2314 commented:
@nbp Ok ready for review! After breaking out all those PRs, it will be nice to merge the last bits of the original!

Ericson2314 commented:
@nbp ping

nbp commented on pkgs/top-level/default.nix:
```diff
@@ -58,8 +69,14 @@ in let
+    inherit lib allPackages system platform crossSystem config;
+  }) stdenv;
+
+  pkgs = allPackages { inherit system stdenv config crossSystem platform; };
```
Note:
This commit changes the dependencies of `stdenv`, and clean-up the stdenv story by removing the `defaultStdenv` attribute as well as the `bootStdenv` argument.

Before, stdenv was provided by all-packages, which was iterating multiple time over the `top-level/default.nix` expression.  This commit changes this relation such that the `stdenv` used for building all the *final* packages depends on the `allPackages` function, which does a fix-point over the set of packages needed for the bootstrap.  And the *final* set of packages depends on the `allPackages` function and on the `stdenv` result.

Ericson2314 replied:
Incorporated into the commit message. Thanks.

nbp commented on pkgs/stdenv/freebsd/default.nix:
```diff
@@ -5,7 +5,7 @@
 
 rec {
-  allPackages = import ../../..;
+  inherit allPackages;
```
nit: this modification should be part of the previous commit.

Ericson2314 replied:
Right. Thanks, I meant to do that and forget.

nbp commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -1,14 +1,14 @@
     config = builtins.removeAttrs config [ "replaceStdenv" ];
-  };
-  vanillaStdenv = (import ../. (args // argClobber // {
-    allPackages = args: allPackages (argClobber // args);
```
Note:
This is fine, because, as the stdenv are now explicitly forced to provide a `config` attribute, by the fact that this commit removes the `config` value from `top-level/default.nix`. This `config` attribute would necessary be the `config` attribute given as argument to the `vanillaStdenv`.  So there is no need to make a custom version of `allPackages` here.

Ericson2314 replied:
Precisely.

nbp commented on pkgs/top-level/platforms.nix:
```diff
@@ -443,4 +443,12 @@ rec {
   };
 
+  chooseDefault = { system }:
+    if system == "armv6l-linux" then raspberrypi
```
nit: rename `chooseDefault` by `selectPlatformBySystem`.  No need to wrap the argument in an attribute set.

nbp commented on pkgs/stdenv/debug.nix:
```diff
@@ -0,0 +1,27 @@
+# This file mimicks top-levevl/default.nix, and is used to provide default
```
typo: `top-levevl`  --> `top-level`.

nbp commented on pkgs/stdenv/native/default.nix:
```diff
@@ -1,4 +1,9 @@
-{ system, allPackages ? import ../../.., config }:
+{ _defaults    ? import ../debug.nix
+, argsResolved ? _defaults.extend (_: _: args)
+, allPackages  ? argsResolved.allPackages
```
I honestly think the code is better without these debug code.
This debug code makes it much more harder to understand what is going on with the argument.

I think if you are an stdenv developer adding a few `builtins.trace` function call is probably the easiest.
As well as making a wrapper file to invoke the right stage of the bootstrap you are looking for.

Ericson2314 replied:
I agree in principle, but here are two headaches:

1. One of the nixpkgs tests is building darwin bootstrap tools and booting a stdenv with it, so I need a decent portion of this complexity just to make an `allPackages` argument suitable for that test to pass.

2. I beleive the goal wrt debugging is less tracing than being able to load a small subset up, potentially with different arguments. That's also why these files are big `rec { .. }` rather than `let .. in ..`.

@copumpkin I think you are the person to ask via both of these, and more broadly what to do with this final commit.

nbp reviewed (changes requested):
Remove the last commit, and fix the nits, and I think this would be ok for merging.

Ericson2314 commented on pkgs/top-level/default.nix:
```diff
@@ -72,75 +57,9 @@ let
   # deterministically inferred the same way.
   nixpkgsFun = newArgs: import ./. (args // newArgs);
```
There should be comments to explain this, but actually I think this is OK. `config` is fine to `include`, but platform actually shouldn't be included because this function is usually called with a non-native `system`, which causes a different `platform` to be inferred. It would be bad to force the native `system`'s inferred `platform`.

Ericson2314 replied:
There should be comments to explain whatever route is chosen, in any event.

Ericson2314 replied:
Ok explained in `HEAD^`.

Ericson2314 commented on pkgs/top-level/default.nix:
```diff
@@ -72,75 +57,9 @@ let
-  # The result is `pkgs' where all the derivations depending on `foo'
-  # will use the new version.
+  # Partially apply some args for building phase pkgs sets
+  pkgs = import ./stage.nix ({
```
Hmm? `stage.nix` makes only a *single* stage--I am using stage in the sense of bootstrapping stage, not "fix-extend" stage, I assume that is the  source of the confusion? I would like better terms to distinguish the two.

Ericson2314 replied:
Ok, I did some `s/phase/stage/` which should help. We already use "phase' for `mkDerivation`, and "stage" for this.

Ericson2314 commented on pkgs/top-level/stage.nix:
```diff
@@ -80,10 +80,9 @@ let
 
   # The complete chain of package set builders, applied from top to bottom
-  toFix = lib.foldl' (lib.flip lib.extends) (self: {}) [
+  toFix = lib.foldl' (lib.flip lib.extends) (self: stdenvDefault) [
```
Sure

Ericson2314 replied:
Doing that

Ericson2314 committed:
top-level: Move default-choosing logic to top-level/platforms.nix

This mirrors stdenv/default.nix

Ericson2314 committed:
top-level: Document why `nixpkgsFun` doesn't close over inferred default arguments

Ericson2314 committed:
top-level: Split some of `pkgs/top-level/default.nix` to `pkgs/top-level/stage.nix`

This is preparation for the latter just building a single stage, and the
former building a package set with the bootstrapped stdenv.

Ericson2314 committed:
top-level: Remove cycles: stdenv calls in top-level but not vice versa

This commit changes the dependencies of stdenv, and clean-up the stdenv
story by removing the `defaultStdenv` attribute as well as the `bootStdenv`
parameter.

Before, the final bootstrapping stage's stdenv was provided by
all-packages, which was iterating multiple times over the
top-level/default.nix expression, and non-final bootstrapping stages'
stdenvs were explicitly specified with the `bootStdenv` parameter.

Now, all stages' stdenvs are specified with the `stdenv` parameter.
For non-final bootstrapping stages, this is a small change---basically just
rename the parameter.
For the final stage, top-level/default.nix takes the chosen stdenv and
makes the final stage with it.

`allPackages` is used to make all bootstrapping stages, final and
non-final alike. It's basically the expression of `stage.nix` (along with a
few partially-applied default arguments)

Note, the make-bootstrap-tools scripts are temporarily broken

Ericson2314 committed:
top-level: Close over fewer arguments for stdenv stages

This makes the flow of data easier to understand. There's little downside
because the args in question are already inspected by the stdenvs.

cross-compiling in particular is simpler because we don't need to worry
about overriding the config closed over by `allPackages`.

Ericson2314 committed:
top-level: turn the screw

 - Non-cross stdenvs are honest and assert that `crossSystem` is null

 - `crossSystem` is a mandatory argument to top-level/stage.nix, just like
   `system` and `platform`

 - Broken default arguments on stdenvs for testing are gone.

 - All stdenvs (but little-used stdenvNix) take the same arguments for easy
   testing.

Ericson2314 committed:
Move up `inherit` binding for consistency

Ericson2314 commented:
@nbp Do you have time to re-review? Basically I went the opposite course of stripping the all the debugging hand-holds. Differences are roughly:
 - No more default arguments for debugging
 - `crossSystem` no longer defaults to `null` per stage and must be explicitly provided (but this is analogous to `system` and `platform`)
 - The first commit now also makes `pkgs/stdenv/default.nix` just the default rather than an attrset of them all (the others were already not exposed by that commit)
 - `top-level/default.nix` can take a custom stdenv now---one could already override it with config but same goes for platform and that's manually specifiable too.

If you want to diff, github.com/Ericson2314/nixpkgs/tree/cross-debug is the old version, and `cross-*` are some other ideas I tossed around. (github's comparing of diverged branches is unfortunately totally broken, so you'd have to use the CLI.0

Ericson2314 committed:
top-level: Allow manually specifying a stdenv, and fix stdenv tests

 - The darwin test can now force the use of the freshly-booted darwin stdenv
 - The linux test now passes enough dummy arguments

This may make debugging harder, if so, check out #20889

Ericson2314 commented:
OK, so its been many months. @nbp told me he doesn't have time to review again for a while, but this is exactly in the spirit of his last review. 

There was only one part of his review I had reservations about, pertaining to my last commit. While removing the debugging aids (default arguments never used normally) does reduce complexity, it also may make debugging harder, and stdenvs are already pain enough. To alleviate that, I rebased my old final commit on top of this and opened  https://github.com/NixOS/nixpkgs/pull/20889 with it. I really want to see this merged, so my hope is that other PR can be merged by anyone who is adversely impacted by the removal of debugging aids in this.

So, my plan is to wait a day or two and then merge this myself. **Speak now, or ~~forever hold your peace~~ merge https://github.com/NixOS/nixpkgs/pull/20889**.
